src/main/java/io/vertx/core/http/impl/pool/Pool.java

```diff
@@ -0,0 +1,315 @@
+/*
+ * Copyright (c) 2011-2014 The original author or authors
+ * ------------------------------------------------------
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Apache License v2.0 which accompanies this distribution.
+ *
+ *     The Eclipse Public License is available at
+ *     http://www.eclipse.org/legal/epl-v10.html
+ *
+ *     The Apache License v2.0 is available at
+ *     http://www.opensource.org/licenses/apache2.0.php
+ *
+ * You may elect to redistribute this code under either of these licenses.
+ */
+package io.vertx.core.http.impl.pool;
+
+import io.netty.channel.Channel;
+import io.vertx.core.Handler;
+import io.vertx.core.http.ConnectionPoolTooBusyException;
+import io.vertx.core.impl.ContextImpl;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import io.vertx.core.spi.metrics.HttpClientMetrics;
+
+import java.util.*;
+import java.util.function.BiConsumer;
+
+/**
+ * The pool is a queue of waiters and a list of connections.
+ *
+ * The pools invariant:
+ * - the pool {@link #capacity} is the sum of the {@link Holder#capacity} of the {@link #available} list
+ * - a connection in the {@link #available} list has its {@code Holder#capacity > 0}
```

Comment:
Why strictly bigger? Can't it be in the `available` dequeue without being borrowed at all?

Reply:
a connection has `capacity == 0` means it cannot accept anymore connection